File: SevenNightClassPCS/oop-cpp/tictactoe1/Board.cpp
#include "Board.h"

#include <string>
using namespace std;

Board::Board() {
	clearBoard();
}

Board::~Board() {
}

string Board::getStringRepr() {
	// In the OO version, change this to a call on getCell
	string ret = string(1,board[0])+"|"+string(1,board[1])+"|"+string(1,board[2])+"\n";
	ret = ret + "-+-+-\n";
	ret = ret + string(1,board[3])+"|"+string(1,board[4])+"|"+string(1,board[5])+"\n";
	ret = ret + "-+-+-\n";
	ret = ret + string(1,board[6])+"|"+string(1,board[7])+"|"+string(1,board[8]);
	return ret;
}

bool Board::checkThreeInARow(char token) {
	if(board[0]==token && board[1]==token && board[2]==token) return true;
	if(board[3]==token && board[4]==token && board[5]==token) return true;
	if(board[6]==token && board[7]==token && board[8]==token) return true;
	if(board[0]==token && board[3]==token && board[6]==token) return true;
	if(board[1]==token && board[4]==token && board[7]==token) return true;
	if(board[2]==token && board[5]==token && board[8]==token) return true;
	if(board[0]==token && board[4]==token && board[8]==token) return true;
	if(board[2]==token && board[4]==token && board[7]==token) return true;
	return false;
}

char Board::getStatus() {
	if(checkThreeInARow('X')) return 'X';
	if(checkThreeInARow('O')) return 'O';
	for(int x=0;x<9;++x) {
		if(board[x]==' ') return ' ';
	}
	return 'C';
}

char Board::getCell(int cell) {
	return board[cell];
}

void Board::makeMove(int cell, char token) {
	board[cell] = token;
}

void Board::clearBoard() {
	for(int x=0;x<9;++x) {
		board[x] = ' ';
	}
}

File: Topics/01_Encapsulation/01_05_TicTacToe_example/python_source/board.py

class Board:
    
    def __init__(self):
        self.clear_board()
        
    def get_string_repr(self):
        ret = self._board[0]+'|'+self._board[1]+'|'+self._board[2]+'\n'
        ret = ret + '-+-+-\n'
        ret = ret + self._board[3]+'|'+self._board[4]+'|'+self._board[5]+'\n'
        ret = ret + '-+-+-\n'
        ret = ret + self._board[6]+'|'+self._board[7]+'|'+self._board[8]
        return ret
    
    # This is for our private use. Outsiders shouldn't need it.
    def _check_for_three_in_a_row(self,token):
        # Horizontal
        if self._board[0] == token and self._board[1] == token and self._board[2] == token:
            return True
        if self._board[3] == token and self._board[4] == token and self._board[5] == token:
            return True
        if self._board[6] == token and self._board[7] == token and self._board[8] == token:
            return True
        # Vertical
        if self._board[0] == token and self._board[3] == token and self._board[6] == token:
            return True
        if self._board[1] == token and self._board[4] == token and self._board[7] == token:
            return True
        if self._board[2] == token and self._board[5] == token and self._board[8] == token:
            return True
        # Diagonals
        if self._board[0] == token and self._board[4] == token and self._board[8] == token:
            return True
        if self._board[2] == token and self._board[4] == token and self._board[6] == token:
            return True
        # If none
        return False

    def get_status(self):
        if self._check_for_three_in_a_row('X'):
            return 'X'
        if self._check_for_three_in_a_row('O'):
            return 'O'    
        for c in self._board:
            if c == ' ':
                return ' '
        return 'C'        

    def make_move(self,cell_number,token):
        # TODO error checking here
        self._board[cell_number] = token
    
    def get_cell(self,cell_number):
        return self._board[cell_number]
    
    def clear_board(self):
        self._board = [' '] * 9
